client/README.md

```diff
@@ -77,6 +77,220 @@ do not apply site-wide.
+using Chrome headless by default.
+
+For testing Vue components, we use the [Vue testing
+utils](https://vue-test-utils.vuejs.org/) to mount individual components in a
+test bed and check them for rendered features.
+
+We use the [babel plugin
+Rewire](https://github.com/speedskater/babel-plugin-rewire) to mock imported ES6
+dependencies during unit-testing. [Sinon](https://sinonjs.org/) is also
+available for more robust stubbing, mocking and spying utilities.
+
+A set of older qUnit tests also exist which will likely be phased-out as the
+code they support is replaced with modern component-based implementations. In
+the meantime, karma runs both the mocha and qunit tests in sequence.
+
+
+## Running the tests
+
+### At Build-Time
+To simply run all the javascript unit tests, you can use make from the root
+directory. This is what happens during a complete client build.
+
+     make client-test
+
+
+### During Development
+
+During client-side development, it is more convenient to have granular testing
+options. The various testing scripts are defined inside package.json within the
+client folder, and are called either with npm (or yarn) as follows:
+
+
+#### Run all tests (mocha and qunit):
+
+     npm run test
+
+#### Watch and rerun all client-side unit-tests every time a source file changes:
+
+     npm run test-watch
+
+#### Run only specified test files when a source file changes:
+
+    npm run test-watch watch-only="Tags/*.test.js"
+
+(The above watch-only parameter will be mapped to the following glob expression "**/Tags/*.test.js")
+
+#### Run qunit legacy tests
+
+    npm run test-qunit
+
+
+## Writing a test file
+
+The karma configs are setup to look at any file ending in "*.test.js". Please
+place your test files inside client/galaxy/scripts folders right next to
+whatever files that they are testing.
+
+```javascript
+// yourtestfile.test.js
+
+import { things } from "./yourtestfile.js";
+
+describe("some module you wrote", () => {
+
+    let transientVariables, serviceInstances, testData;
+
+    beforeEach(() => {
+        // setup your code (if necessary)
+    })
+
+    afterEach(() => {
+        // teardown your code (so it doesn't ruin the next test)
+    })
+
+    it("should do something or other", () => {
+        assert(somethingThatShouldTrue, "explain what went wrong");
+        expect(something).to.eq(true);
+    })
+})
+```
+
+
+## Testing ~~Suggestions~~ Obligations
+
+
+### Clearly document the intent of your test
+
+Please remember that these tests are not *for* you. They're for the people who
+come after you. It will be a lot easier to modify, repair and upgrade your code
+if they can figure out what you were originally hoping to accomplish.
+
+Add a couple comments. Use variable names that mean something. Use the assertion
+failure description parameters when possible. I'm not sure what your mom told
+you, but your code is almost definitely not self-documenting.
```

Comment:
I think we should just drop the your mom statement and convey the same message.

Reply:
Removed! I didn't trust your mom anyway.